src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/OrdinalStringFrozenSet.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+
+namespace System.Collections.Frozen
+{
+    /// <summary>Provides a frozen set optimized for ordinal (case-sensitive or case-insensitive) lookup of strings.</summary>
+    internal sealed class OrdinalStringFrozenSet : FrozenSetInternalBase<string, OrdinalStringFrozenSet.GSW>
+    {
+        private readonly FrozenHashTable _hashTable;
+        private readonly string[] _items;
+        private readonly StringComparerBase _partialComparer;
+        private readonly int _minimumLength;
+        private readonly int _maximumLengthDiff;
+
+        internal OrdinalStringFrozenSet(HashSet<string> source, IEqualityComparer<string> comparer) :
+            base(comparer)
+        {
+            Debug.Assert(source.Count != 0);
+            Debug.Assert(comparer == EqualityComparer<string>.Default || comparer == StringComparer.Ordinal || comparer == StringComparer.OrdinalIgnoreCase);
+
+            string[] entries = new string[source.Count];
+            source.CopyTo(entries);
+
+            _items = new string[entries.Length];
+
+            _partialComparer = ComparerPicker.Pick(
+                entries,
+                ignoreCase: ReferenceEquals(comparer, StringComparer.OrdinalIgnoreCase),
+                out _minimumLength,
+                out _maximumLengthDiff);
+
+            _hashTable = FrozenHashTable.Create(
+                entries,
+                _partialComparer.GetHashCode,
+                (index, item) => _items[index] = item);
+        }
+
+        /// <inheritdoc />
+        private protected override ImmutableArray<string> ItemsCore => new ImmutableArray<string>(_items);
+
+        /// <inheritdoc />
+        private protected override Enumerator GetEnumeratorCore() => new Enumerator(_items);
+
+        /// <inheritdoc />
+        private protected override int CountCore => _hashTable.Count;
+
+        /// <inheritdoc />
+        private protected override int FindItemIndex(string item)
+        {
+            if (item is not null && // this implementation won't be used for null values
```

Comment:
Should this be throwing instead when it's null? The signature says a non-null string...

Reply:
A developer could use a T? in which case null is part of the valid domain, so we need to accomodate nulls being passed in.  This is a hole in the nullable reference types analysis.  I can update the type here to be string? to better highlight it.